### PR TITLE
Support sync start from certain evm height

### LINKED ISSUE #151: Investigate why the mechanism to start EOS EVM Node from a specific block doesn’t work
Also add tests for this feature since they don't currently exist

It is not in scope to currently change the interface. If we build a case for this, we can introduce a follow-on issue.

### PR DESCRIPTION
**Pull request by yarkinwho**
Remove support for start from certain eos id functionality.

Resolve #151

### CODE REVIEW

File: src/ship_receiver_plugin.hpp [hunk 1]
```diff
@@ -21,7 +21,7 @@ class ship_receiver_plugin : public appbase::plugin<ship_receiver_plugin> {
       void plugin_startup();
       void plugin_shutdown();
 
-      std::optional<channels::native_block> get_start_from_block();
+      uint64_t get_start_from_canonical_height();
```
Comment (elmato) [inline src/ship_receiver_plugin.hpp:24]: same here

File: src/engine_plugin.cpp [hunk 2]
```diff
@@ -136,13 +136,16 @@ class engine_plugin_impl : std::enable_shared_from_this<engine_plugin_impl> {
          return silkworm::db::read_canonical_header(txn, head_num);
       }
 
-      std::optional<silkworm::Block> get_head_block() {
+      std::optional<silkworm::Block> get_head_block(uint64_t override_height) {
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply src/engine_plugin.cpp]: done

File: src/ship_receiver_plugin.cpp [hunk 3]
```diff
@@ -29,21 +29,19 @@ class ship_receiver_plugin_impl : std::enable_shared_from_this<ship_receiver_plu
       using native_block_t = channels::native_block;
       static constexpr eosio::name pushtx = eosio::name("pushtx");
 
-      void init(std::string h, std::string p, eosio::name ca,
-                std::optional<eosio::checksum256> start_block_id, int64_t start_block_timestamp,
+      void init(std::string h, std::string p, eosio::name ca, uint64_t input_start_height,
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply src/ship_receiver_plugin.cpp]: done

File: src/ship_receiver_plugin.hpp (hunk 1 above)
Comment (yarkinwho) [reply src/ship_receiver_plugin.hpp]: done

**Comment by yarkinwho** [comment]
please note that the logic for get_head_block is also changed a little bit:

Now we will not try fetch canonical header at all if the override option is present. <- The overriding is meant to fix the db anyway, so we should give it a try.

File: src/engine_plugin.cpp [hunk 4]
```diff
@@ -136,13 +136,22 @@ class engine_plugin_impl : std::enable_shared_from_this<engine_plugin_impl> {
          return silkworm::db::read_canonical_header(txn, head_num);
       }
 
-      std::optional<silkworm::Block> get_head_block() {
-         auto header = get_head_canonical_header();
-         if(!header) return {};
+      std::optional<silkworm::Block> get_canonical_block_at_height(const std::optional<uint64_t>& override_height) {
```
Comment (elmato) [inline src/engine_plugin.cpp:139]: override_height => height

Comment (elmato) [reply src/engine_plugin.cpp:139]: Also for small types (uint64_t) we can pass std::optional by value.
```c++
std::optional<silkworm::Block> get_canonical_block_at_height(std::optional<uint64_t> height)
```

File: src/engine_plugin.cpp [hunk 5]
```diff
@@ -211,8 +220,8 @@ std::optional<silkworm::BlockHeader> engine_plugin::get_head_canonical_header()
    return my->get_head_canonical_header();
 }
 
-std::optional<silkworm::Block> engine_plugin::get_head_block() {
-   return my->get_head_block();
+std::optional<silkworm::Block> engine_plugin::get_canonical_block_at_height(const std::optional<uint64_t>& override_height) {
+   return my->get_canonical_block_at_height(override_height);
```
Comment (elmato) [inline src/engine_plugin.cpp:224]: same as above

File: src/engine_plugin.hpp [hunk 6]
```diff
@@ -22,7 +22,7 @@ class engine_plugin : public appbase::plugin<engine_plugin> {
       uint32_t    get_threads() const;
       std::string get_chain_data_dir() const;
       std::optional<silkworm::BlockHeader> get_head_canonical_header();
-      std::optional<silkworm::Block> get_head_block();
+      std::optional<silkworm::Block> get_canonical_block_at_height(const std::optional<uint64_t>& override_height);
```
Comment (elmato) [inline src/engine_plugin.hpp:25]: same as above

File: src/ship_receiver_plugin.cpp [hunk 7]
```diff
@@ -29,21 +29,19 @@ class ship_receiver_plugin_impl : std::enable_shared_from_this<ship_receiver_plu
       using native_block_t = channels::native_block;
       static constexpr eosio::name pushtx = eosio::name("pushtx");
 
-      void init(std::string h, std::string p, eosio::name ca,
-                std::optional<eosio::checksum256> start_block_id, int64_t start_block_timestamp,
+      void init(std::string h, std::string p, eosio::name ca, const std::optional<uint64_t>& input_start_height,
```
Comment (elmato) [inline src/ship_receiver_plugin.cpp:32]: same as above

File: src/ship_receiver_plugin.cpp [hunk 8]
```diff
@@ -481,6 +468,7 @@ void ship_receiver_plugin::plugin_shutdown() {
    SILK_INFO << "Shutdown SHiP Receiver";
 }
 
-std::optional<channels::native_block> ship_receiver_plugin::get_start_from_block() {
-   return my->get_start_from_block();
+const std::optional<uint64_t>& ship_receiver_plugin::get_start_from_canonical_height() {
```
Comment (elmato) [inline src/ship_receiver_plugin.cpp:471]: std::optional<uint64_t>

File: src/ship_receiver_plugin.hpp [hunk 9]
```diff
@@ -21,7 +21,7 @@ class ship_receiver_plugin : public appbase::plugin<ship_receiver_plugin> {
       void plugin_startup();
       void plugin_shutdown();
 
-      std::optional<channels::native_block> get_start_from_block();
+      const std::optional<uint64_t>& get_start_from_canonical_height();
```
Comment (elmato) [inline src/ship_receiver_plugin.hpp:24]: same as above

File: src/block_conversion_plugin.cpp [hunk 10]
```diff
@@ -46,7 +46,8 @@ class block_conversion_plugin_impl : std::enable_shared_from_this<block_conversi
       }
 
       void load_head() {
-         auto head_block = appbase::app().get_plugin<engine_plugin>().get_head_block();
+         const std::optional<uint64_t>& start_from_canonical_height = appbase::app().get_plugin<ship_receiver_plugin>().get_start_from_canonical_height();
```
Comment (elmato) [inline src/block_conversion_plugin.cpp:49]: `auto` can be used here

Comment (yarkinwho) [reply src/block_conversion_plugin.cpp]: done

File: src/ship_receiver_plugin.hpp (hunk 9 above)
Comment (yarkinwho) [reply src/ship_receiver_plugin.hpp]: done

File: src/ship_receiver_plugin.cpp (hunk 8 above)
Comment (yarkinwho) [reply src/ship_receiver_plugin.cpp]: done

File: src/ship_receiver_plugin.cpp [hunk 11]
```diff
@@ -392,15 +390,9 @@ class ship_receiver_plugin_impl : std::enable_shared_from_this<ship_receiver_plu
...
-      std::optional<channels::native_block> get_start_from_block() {
-         if( !start_from_block_id ) return {};
-
-         channels::native_block nb{};
-         nb.id = *start_from_block_id;
-         nb.timestamp = start_from_block_timestamp;
-         nb.block_num = utils::to_block_num(*start_from_block_id);
-         return nb;
-      }
+   const std::optional<uint64_t>& get_start_from_canonical_height() {
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply src/ship_receiver_plugin.cpp]: done

File: src/ship_receiver_plugin.cpp [hunk 12]
```diff
@@ -353,27 +351,27 @@ class ship_receiver_plugin_impl : std::enable_shared_from_this<ship_receiver_plu
...
-            auto head_header = appbase::app().get_plugin<engine_plugin>().get_head_canonical_header();
-            if (!head_header) {
-               sys::error("Unable to read canonical header");
+            // Only take care of canonical header and input options when it's initial sync.
+            if (start_from_canonical_height) {
+               SILK_INFO << "Override head height with"
+                        << "#" << *start_from_canonical_height;
+            }
+
+            auto head_block = appbase::app().get_plugin<engine_plugin>().get_canonical_block_at_height(start_from_canonical_height);
```
[the remark this answers is not in the document]
Comment (yarkinwho) [reply src/ship_receiver_plugin.cpp]: done

File: src/ship_receiver_plugin.cpp (hunk 7 above)
Comment (yarkinwho) [reply src/ship_receiver_plugin.cpp]: done

File: src/engine_plugin.hpp (hunk 6 above)
Comment (yarkinwho) [reply src/engine_plugin.hpp]: done

File: src/engine_plugin.cpp (hunk 4 above)
Comment (yarkinwho) [reply src/engine_plugin.cpp]: done

**Review by elmato — APPROVED** [review]

**Pull request merged**